### PR TITLE
Add Mochi-1 full-res input helpers and transformer shard specs

### PR DESCRIPTION
**Pull request by kamalrajkannan78**
### Ticket

- https://github.com/tenstorrent/tt-xla/issues/4640

### Problem description

- The Mochi loader only exposed a small synthetic input shape (`2×12×12`) and had no SPMD shard specs for the transformer, which needs to be extended to bring up Mochi-1 at its original resolution (480×848, 84 frames, ~10B-param transformer) 

### What's changed

- **`utils.py`**: full-resolution input helpers (`load_{transformer,vae_decoder,text_encoder}_inputs_full_res`) driven by `ORIGINAL_LATENT_*` / `ORIGINAL_MAX_SEQ_LEN` constants matching Mochi-1's real 480×848 / 84-frame shape.

- **`utils.py`**: `MESH_SHAPES`, `MESH_NAMES`, and `shard_transformer_specs` for `MochiTransformer3DModel` — column-parallel Q/K/V + FF-up, row-parallel out-projs + FF-down with replicated biases; `patch_embed` / `proj_out` / norms run replicated; handles `context_pre_only` and `add_q_proj is None`.

- **`utils.py`**: `shard_vae_decoder_specs` for `MochiDecoder3D` — Megatron column–row sharding on each ResBlock's two convs (`conv1` column-parallel, `conv2` row-parallel with replicated bias) and channel-sharded `norm2`; `norm1`, top-level `conv_in`, `proj_out`, and per-up-block `proj` left replicated.

- **`loader.py`**: exposes the above via `get_mesh_config()` and `load_shard_spec()`. Text encoder stays single-chip; transformer and VAE decoder use the same mesh shapes. Partition specs use `None` for non-sharded dims (not `"batch"`), so they're correct on any mesh shape, not just `(1, N)`.

### Checklist
- [x] Verified the changes through local testing in WH

### Logs

- [may12_mochi_encoder.log](https://github.com/user-attachments/files/27654716/may12_mochi_encoder.log)
- [may23_mochi_transformer_8_chips.log](https://github.com/user-attachments/files/28176543/may23_mochi_transformer_8_chips_re.log)
- [may23_mochi_vae_decoder_8_chips.log](https://github.com/user-attachments/files/28176546/may23_mochi_vae_decoder_8_chips_re.log)

### CODE REVIEW

**Review by saiarthiraguram — APPROVED** [review]

File: mochi/pytorch/src/utils.py [hunk 1]
```diff
@@ -274,3 +274,136 @@ def load_text_encoder_inputs(dtype: torch.dtype = torch.bfloat16) -> dict:
...
+        # Q/K/V text stream (column-parallel, added_proj_bias=False in Mochi).
+        if attn.add_q_proj is not None:
+            specs[attn.add_q_proj.weight] = ("model", "batch")
+        specs[attn.add_k_proj.weight] = ("model", "batch")
+        specs[attn.add_v_proj.weight] = ("model", "batch")
+
+        # Output projections (row-parallel, bias=True by default).
+        specs[attn.to_out[0].weight] = ("batch", "model")
+        if attn.to_out[0].bias is not None:
+            specs[attn.to_out[0].bias] = ("batch",)
```
Comment (vkovinicTT) [inline mochi/pytorch/src/utils.py]: Change `specs[attn.to_out[0].bias]` from `("batch",)` to `(None,)`. Weight stays as `("batch", "model")`(row-parallel along `"model"`, correct): paired with column-parallel to_q/to_k/to_v, this gives the standard Megatron column→SDPA→row sandwich with a single all-reduce at the end. Since that all-reduce already produces the full output on every chip, the `bias` must be replicated so each chip adds the full vector — sharding it along `"batch"` only worked on n300 because that axis is size 1.

File: mochi/pytorch/src/utils.py [hunk 2]
```diff
@@ -274,3 +274,136 @@ def load_text_encoder_inputs(dtype: torch.dtype = torch.bfloat16) -> dict:
...
+        if not attn.context_pre_only and hasattr(attn, "to_add_out"):
+            specs[attn.to_add_out.weight] = ("batch", "model")
+            if attn.to_add_out.bias is not None:
+                specs[attn.to_add_out.bias] = ("batch",)
```
Comment (vkovinicTT) [inline mochi/pytorch/src/utils.py]: Change `specs[attn.to_add_out.bias]` from `("batch",)` to `(None,)`. Weight stays as `("batch", "model")` (row-parallel along `"model"`, correct): same column→SDPA→row sandwich, one all-reduce, same bias-replication argument.

File: mochi/pytorch/src/utils.py [hunk 3]
```diff
@@ -274,3 +274,136 @@ def load_text_encoder_inputs(dtype: torch.dtype = torch.bfloat16) -> dict:
...
+    # Patch embedding (PatchEmbed wraps a Conv2d at .proj).
+    if hasattr(transformer.patch_embed, "proj"):
+        specs[transformer.patch_embed.proj.weight] = (
+            "batch",
+            None,
+            None,
+            None,
+        )
+        if transformer.patch_embed.proj.bias is not None:
+            specs[transformer.patch_embed.proj.bias] = ("batch",)
```
Comment (vkovinicTT) [inline mochi/pytorch/src/utils.py]: Remove both `patch_embed.proj.weight` and `patch_embed.proj.bias` entries entirely (`Conv2d` runs replicated).

File: mochi/pytorch/src/utils.py [hunk 4]
```diff
@@ -274,3 +274,136 @@ def load_text_encoder_inputs(dtype: torch.dtype = torch.bfloat16) -> dict:
...
+        specs[block.ff.net[0].proj.weight] = ("model", "batch")
+        specs[block.ff.net[2].weight] = ("batch", "model")
+        if block.ff_context is not None:
+            specs[block.ff_context.net[0].proj.weight] = ("model", "batch")
+            specs[block.ff_context.net[2].weight] = ("batch", "model")
+
+    # Output head.
+    specs[transformer.proj_out.weight] = (None, "batch")
+    if transformer.proj_out.bias is not None:
+        specs[transformer.proj_out.bias] = (None,)
```
Comment (vkovinicTT) [inline mochi/pytorch/src/utils.py]: Final Linear runs replicated as well.

File: mochi/pytorch/src/utils.py [hunk 5]
```diff
@@ -274,3 +274,183 @@ def load_text_encoder_inputs(dtype: torch.dtype = torch.bfloat16) -> dict:
...
+    return [input_ids, attention_mask]
+
+
+# ============================================================================
+# SPMD shard specifications (transformer)
+# ============================================================================
+
+# (batch, model) mesh shapes by device count. n300 = 2 chips.
+MESH_SHAPES = {1: (1, 1), 2: (1, 2), 4: (1, 4), 8: (1, 8), 32: (8, 4)}
+MESH_NAMES = ("batch", "model")
```
Comment (vkovinicTT) [inline mochi/pytorch/src/utils.py]: Please pay attention whether maths checks right for all mesh shapes, since your's only worked because you were running it on n300 where first shape size is 1 and it efectively replicated things, but if you were to run it on the galaxy `(8,4)` for example, it wouldn't work. Please be careful about this.

**Review by vkovinicTT — CHANGES_REQUESTED** [review]
Also add sharding for VAE, you can take a look at this [issue](https://github.com/tenstorrent/tt-xla/issues/4252) which has sharding on branch: `vkovinic/mochi`.

**Comment by kamalrajkannan78** [comment]
> Also add sharding for VAE, you can take a look at this [issue](https://github.com/tenstorrent/tt-xla/issues/4252) which has sharding on branch: `vkovinic/mochi`.

`shard_vae_decoder_specs` is already in this PR — ported from your _shard_decoder_weights on [vkovinic/mochi](https://github.com/tenstorrent/tt-xla/compare/main...vkovinic/mochi).

**Review by vkovinicTT — APPROVED** [review]

**Comment by vkovinicTT** [comment]
before merging, please provide new logs, since previous logs were for different sharding.

**Comment by kamalrajkannan78** [comment]
> before merging, please provide new logs, since previous logs were for different sharding.

Tested the decoder and transformer on top of Friday’s main branch and updated the logs in the description.
